## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [63]:
import nsfg
df = nsfg.ReadFemPreg()

In [64]:
import thinkstats2
import pandas as pd
import math

cor = pd.DataFrame(index = df.columns)

for feature in df.columns:
    cor.loc[feature, 'pearson'] = abs(thinkstats2.Corr(df[feature], df.prglngth))
    cor.loc[feature, 'spearman'] = abs(thinkstats2.SpearmanCorr(df[feature], df.prglngth))
    
print cor.sort_values(by='pearson', ascending=False)

              pearson  spearman
prglngth     1.000000  1.000000
outcome      0.738494  0.789606
parity       0.272541  0.255017
wantresp     0.198823  0.203856
oldwantr     0.198710  0.203884
oldwantp     0.185573  0.171102
wantpart     0.183213  0.169948
hpwnold      0.163656  0.146468
pregnum      0.150141  0.126270
fmarcon5     0.149385  0.137865
datend_i     0.132925  0.123302
datecon_i    0.132414  0.122378
agecon_i     0.123470  0.113589
agepreg_i    0.120842  0.114430
poverty      0.111886  0.072126
rmarital     0.087367  0.092605
fmarital     0.086664  0.090788
pncarewk_i   0.085949  0.082594
educat       0.075929  0.048979
hieduc       0.073817  0.043487
pmarpreg_i   0.072958  0.069937
rcurpreg     0.072415  0.082847
hispanic     0.069856  0.045449
wantpart_i   0.062641  0.058936
learnprg_i   0.062127  0.072361
birthord_i   0.061499  0.057035
outcome_i    0.061499  0.057035
parity_i     0.060742  0.061266
prglngth_i   0.059447  0.093909
paydeliv_i   0.059436  0.053797
...     

In [85]:
from sklearn import linear_model
clf = linear_model.LinearRegression()

prglngth = df.prglngth.reshape((len(df), 1))

for feature in df.columns:
    if (feature == 'caseid'):
        continue
    
    frame = df.dropna(subset=[feature])
    if (len(frame[feature]) == 0):
        continue

    data = frame[feature].reshape((len(frame), 1))
    clf.fit(data, frame.prglngth)
    y_pred = clf.predict(data)
    
    r2 = thinkstats2.CoefDetermination(frame.prglngth, y_pred)
    if (r2 > 0):
        p = math.sqrt(r2)
    
    cor.loc[feature, 'r2'] = r2
    cor.loc[feature, 'p'] = p
    
print cor

                  pearson  spearman            r2             p
caseid           0.012622  0.013633  9.998407e-01  9.999203e-01
pregordr         0.057760  0.037535  9.966637e-01  9.983305e-01
howpreg_n             NaN       NaN  7.652855e-01  8.748060e-01
howpreg_p             NaN       NaN  9.851035e-01  9.925238e-01
moscurrp              NaN       NaN  6.432386e-04  2.536215e-02
nowprgdk              NaN       NaN  7.771561e-16  2.787752e-08
pregend1              NaN       NaN  2.539247e-01  5.039094e-01
pregend2              NaN       NaN  9.004885e-01  9.489407e-01
nbrnaliv              NaN       NaN  9.849686e-01  9.924558e-01
multbrth              NaN       NaN  9.263940e-01  9.624936e-01
cmotpreg              NaN       NaN  9.999953e-01  9.999976e-01
prgoutcome            NaN       NaN  2.520795e-01  5.020752e-01
cmprgend              NaN       NaN  9.859182e-01  9.929341e-01
flgdkmo1              NaN       NaN  9.999633e-01  9.999817e-01
cmprgbeg              NaN       NaN  9.9

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

I wasn't sure if this question was asking us to do analysis on the data to answer the question, or just based on what we've looked at in the data so far.  So I decided to go ahead and see what I could learn from the data


In [16]:
import nsfg
df = nsfg.ReadFemPreg()

c = df.corr().abs()

print c.prglngth.sort_values(ascending=False)

lobthwgt3        1.000000
ltrimest         1.000000
prglngth         1.000000
nowprgdk         1.000000
dk2gest          1.000000
moscurrp         0.999678
wksgest          0.998124
mosgest          0.944177
prgoutcome       0.864824
pregend1         0.863757
outcome          0.738494
alivenow3        0.661277
legagree2        0.627456
dk3gest          0.569174
weeksdk          0.552055
parenend2        0.524049
getprena         0.501113
lastage3         0.499266
totalwgt_lb      0.494013
birthwgt_lb      0.485165
howpreg_n        0.484473
lbw1             0.479601
pncarewk         0.468033
pnctrim          0.457940
babysex3         0.448897
livehere3        0.418246
birthwgt_lb3     0.409182
alivenow         0.402706
matchfound3      0.337903
lpnctri          0.329442
                   ...   
cmintfinop       0.000746
workborn         0.000650
cmprgbeg         0.000429
paybirth3        0.000000
dk1gest               NaN
bpa_bdscheck2         NaN
bpa_bdscheck3         NaN
cmkidlft3   

I tried to use pandas to find possible correlations between the length of the pregnancy and other variables, but it ultimately didn't prove to be very helpful.  The variables I was hoping to see having an impact (ones that we could easily measure from a coworker) like mother's age or birth order didn't show up as high-correlation.  The variables that did show up are often data that comes after birth, like birthweight (makes sense that this is tied to length, preemie babies have less time to grow), or the outcome of the birth, or obvious things like other measure of pregnancy length.

In [20]:
import chap01soln
import statsmodels.formula.api as smf

df = df[df.prglngth>30]
resp = chap01soln.ReadFemResp()
resp.index = resp.caseid
join = df.join(resp, on='caseid', rsuffix='_r')

t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7:
            continue
        formula = 'prglngth ~ agepreg + ' + str(name)
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    t.append((results.rsquared, name))
    
t.sort(reverse=True)
for mse, name in t[:30]:
    print name, mse

prglngth 1.0
wksgest 0.810466988405
agecon 0.781350418865
totalwgt_lb 0.125503308033
birthwgt_lb 0.120679453558
lbw1 0.103838874743
mosgest 0.0982884576401
prglngth_i 0.0215849926795
pregend1 0.00843280219459
outcome 0.00787817147656
prgoutcome 0.00784661257936
canhaver 0.00675998334427
datcon01_i 0.00590185285858
con1mar1_i 0.00559512377641
nbrnaliv 0.00458014116394
mar1con1_i 0.00328669224473
anynurse 0.00291848398136
bfeedwks 0.00256701215878
evuseint 0.0021679382829
rmarout11_i 0.00216747068093
marout11_i 0.00216747068093
marcon11_i 0.00216747068093
datend02_i 0.00207103036159
datcon02_i 0.00207103036159
agecon02_i 0.00202887790055
ageprg02_i 0.00199128042069
paydu 0.00198269210289
fmarcon5_i 0.00194217103909
cmlastlb_r 0.00190168382206
cmlastlb 0.00190168382206


This time, I used the statsmodel approach as shown in the book, but again, the results were not terribly helpful.  Obviously, it correlates with prglngth, and a lot of the variables, like outcome or totalwgt_lb will only be relevant after the actual birth.

Age at conception is easily determined, but is obviously correlated with age of the mother now, so probably doesn't add any new information.  A variable like canhaver (ability to have another baby) might be another that we could figure oyuut pre-brith from the coworker, but at this point the R-squared value is under 1%, which means it really isn't having much of an impact anyways.  

If I were actually doing this for my office, I would just collect as many variables from my coworker as possible, such as age, birth order of child, race, babysex, did they recieve prenatal care, or any other variables that show up in the dataset.  Then, I could make a training model based only on the variables I have, in order to attempt to predict pregnancy length.


## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

I would consider using the following factors - 

- Age - easy variable to gather, seems like it had some significance
- Height/Weight? - may hold some significance
- First child or not
- Race - this seemed to hold some significance in the baby's weight
- Gender of baby - questionable whether this could be discerned at that point
- Number of babies born - again, questionable if this could actually be discerned
- Income & related variables - income, as mentionted in ThinkStats, generally has a pretty sizeable effect on many other aspect's of people's lives, and any variability due to income would likely account for these 

There are many other variables that could be learned (e.g. diet and health of the mom), but I'd bet a lot of these are also dependent on some combination of the variables mentioned above, and their predictive power would be insignificant if included. 

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

First, I want to load in the data and just restrict the database right off the bat. I'll also use this time to drop my nan values. 

In [82]:
births = nsfg.ReadFemPreg()
live = births[(births.outcome == 1) & (births.prglngth > 30)]
resp = chap01soln.ReadFemResp()


In [83]:
join = live.join(resp, on='caseid', rsuffix='_r')


In [84]:
t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7:
            continue
        formula = ('prglngth ~ ' + name).encode('ascii')
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    t.append((results.rsquared, name))

In [85]:
t.sort(reverse=True)
for mse, name in t[:30]:
    print(name, mse)

(u'prglngth', 1.0)
(u'wksgest', 0.80624341161392343)
('totalwgt_lb', 0.12445743148120236)
(u'birthwgt_lb', 0.11977307804917292)
(u'lbw1', 0.10372542204583624)
(u'mosgest', 0.095624319895927679)
(u'prglngth_i', 0.0220537757964685)
(u'nbrnaliv', 0.0045775657855365859)
(u'mardat02_i', 0.003084214232312199)
(u'wantrp07_i', 0.0028805878918768402)
(u'oldwr07_i', 0.0028805878918768402)
(u'oldwp07_i', 0.0028805878918768402)
(u'parts12', 0.0027898760463058725)
(u'rmarout07_i', 0.0027500058046551201)
(u'condomr_i', 0.0024693444168197853)
(u'anynurse', 0.0024520248837134329)
(u'bfeedwks', 0.0023691839446664531)
(u'mon12prt', 0.0023356623983744607)
(u'pregend1', 0.0022493894338005971)
(u'intr_ec3', 0.0020549332842115797)
(u'cmlastlb', 0.0020431424422021616)
(u'warm', 0.0020189345316214968)
(u'fmarcon5_i', 0.0019681593242552031)
(u'evuseint', 0.0018917527758629538)
(u'agecon07_i', 0.0018117332152514098)
(u'marcon05_i', 0.0017807675795284972)
(u'knowfp', 0.0017743566721165616)
(u'sexever_i', 0.00175

Okay, so this predicts the weeks of gestation for all the variables. Now, let's only use the variables that we'd have access to.

In [86]:
joinFiltered = join[['prglngth','pregordr','agepreg','marstat','hisp',
                     'roscnt','goschol','higrade','havedip','havedeg',
                     'numpregs','cntry00','brnout','evwrk6mo','wrk12mos',
                     'educat','race','numfmhh']]

In [87]:
tFiltered = []
for name in joinFiltered.columns:
    try:
        if join[name].var() < 1e-7:
            continue
        formula = ('prglngth ~ ' + name).encode('ascii')
        model = smf.ols(formula, data=joinFiltered)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    tFiltered.append((results.rsquared, name))

In [88]:
tFiltered.sort(reverse=True)
for mse, name in tFiltered[:30]:
    print(name, mse)

(u'prglngth', 1.0)
(u'pregordr', 0.00062224148606715435)
(u'educat', 0.00059068244033599893)
(u'marstat', 0.00055259939593688134)
(u'numpregs', 0.00039871360360510533)
(u'havedip', 0.00035383982253922586)
(u'race', 0.00031440218809630771)
(u'numfmhh', 0.00027737775283898092)
(u'hisp', 0.00012628508855627718)
(u'roscnt', 0.0001200107015281171)
(u'higrade', 4.7422830185106513e-05)
(u'wrk12mos', 3.6221957431581409e-05)
(u'brnout', 2.7050223118885164e-05)
(u'evwrk6mo', 2.2877588194858411e-05)
(u'agepreg', 1.7603962211398816e-05)
(u'goschol', 1.984423825263093e-07)


So, it doesn't seem like there is anything that's super related to the birth day in the columns that I looked through. Other than, of course, the length of the pregnancy. It seems like the sorts of things that the people at an office would know don't really predict when the baby will be born. 

There might well be different things that I could use to predict the date of birth, but I spent about half an hour looking through the codebook for columns that coworkers might now, and these are the ones I found. Perhaps I missed a different method of selecting column names that might have been faster. 

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [36]:
live = df[df.outcome == 1]
live = live[live.prglngth > 30]

resp = chap01soln.ReadFemResp()
resp.index = resp.caseid
join = live.join(resp, on='caseid', rsuffix='_r')

In [37]:
t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7:
            continue
        formula = 'prglngth ~ agecon + ' + str(name)
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    t.append((results.rsquared, name))

In [38]:
t.sort(reverse=True)
for mse, name in t[:10]:
    print(name, mse)

(u'prglngth', 1.0)
(u'wksgest', 0.80647560229554416)
(u'agepreg', 0.77722780709056005)
('totalwgt_lb', 0.12589504948051811)
(u'birthwgt_lb', 0.12104803951300569)
(u'lbw1', 0.10399010733882053)
(u'mosgest', 0.095653948252776844)
(u'prglngth_i', 0.022702898461402943)
(u'canhaver', 0.0060528525930908517)
(u'datcon01_i', 0.006021881275759422)


The variables above are the ones that I would look at first. In particular, 'wksgest', 'agepreg', and 'prglngth_i', if it possible to get that information. If not, then I would increase the number of variables: 

In [39]:
for mse, name in t[:30]:
    print(name, mse)

(u'prglngth', 1.0)
(u'wksgest', 0.80647560229554416)
(u'agepreg', 0.77722780709056005)
('totalwgt_lb', 0.12589504948051811)
(u'birthwgt_lb', 0.12104803951300569)
(u'lbw1', 0.10399010733882053)
(u'mosgest', 0.095653948252776844)
(u'prglngth_i', 0.022702898461402943)
(u'canhaver', 0.0060528525930908517)
(u'datcon01_i', 0.006021881275759422)
(u'con1mar1_i', 0.0057470796692380421)
(u'nbrnaliv', 0.0046402822824620493)
(u'mar1con1_i', 0.0032946548980058443)
(u'anynurse', 0.0032494403967081587)
(u'bfeedwks', 0.0027926146173661293)
(u'agebaby1', 0.0027862747948664834)
(u'rmarout11_i', 0.0023613798588607571)
(u'marout11_i', 0.0023613798588607571)
(u'marcon11_i', 0.0023613798588607571)
(u'paydu', 0.0023452835507742353)
(u'pregend1', 0.0022912806602890523)
(u'datend02_i', 0.0021324893643125398)
(u'datcon02_i', 0.0021324893643125398)
(u'cmlastlb_r', 0.0021148239919197565)
(u'cmlastlb', 0.0021148239919197565)
(u'agecon02_i', 0.0021129060195237415)
(u'ageprg02_i', 0.0020748312204004193)
(u'fmarcon5_

If the top few variables are not available, then I would refer to this list. 

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [13]:
import nsfg

df = nsfg.ReadFemPreg()
print df.columns

Index([         u'caseid',        u'pregordr',       u'howpreg_n',
             u'howpreg_p',        u'moscurrp',        u'nowprgdk',
              u'pregend1',        u'pregend2',        u'nbrnaliv',
              u'multbrth',
       ...
            u'laborfor_i',      u'religion_i',         u'metro_i',
               u'basewgt', u'adj_mod_basewgt',        u'finalwgt',
                u'secu_p',            u'sest',         u'cmintvw',
           u'totalwgt_lb'],
      dtype='object', length=244)


In [15]:
df = df[df.prglngth>30]
resp = chap01soln.ReadFemResp()
resp.index = resp.caseid
join = df.join(resp, on='caseid', rsuffix='_r')

t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7:
            continue
        formula = 'prglngth ~ agepreg + ' + str(name)
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    t.append((results.rsquared, name))
    
t.sort(reverse=True)
for mse, name in t[:30]:
    print name, mse

prglngth 1.0
wksgest 0.810466988405
agecon 0.781350418865
totalwgt_lb 0.125503308033
birthwgt_lb 0.120679453558
lbw1 0.103838874743
mosgest 0.0982884576401
prglngth_i 0.0215849926795
pregend1 0.00843280219459
outcome 0.00787817147656
prgoutcome 0.00784661257936
canhaver 0.00675998334427
datcon01_i 0.00590185285858
con1mar1_i 0.00559512377641
nbrnaliv 0.00458014116394
mar1con1_i 0.00328669224473
anynurse 0.00291848398136
bfeedwks 0.00256701215878
evuseint 0.0021679382829
rmarout11_i 0.00216747068093
marout11_i 0.00216747068093
marcon11_i 0.00216747068093
datend02_i 0.00207103036159
datcon02_i 0.00207103036159
agecon02_i 0.00202887790055
ageprg02_i 0.00199128042069
paydu 0.00198269210289
fmarcon5_i 0.00194217103909
cmlastlb_r 0.00190168382206
cmlastlb 0.00190168382206


## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [3]:
print "You could look at the age of the pregnant woman, which would probably be the most definitive indicator. Also, the"
print "child number could be important."

You could look at the age of the pregnant woman, which would probably be the most definitive indicator. Also, the
child number could be important.


## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [33]:
data = nsfg.ReadFemPreg()
data = data[(data.prglngth > 30) & (data.outcome == 1)]
resp = chap01soln.ReadFemResp()
resp.index = resp.caseid
joined = data.join(resp, on='caseid', rsuffix='_r')

joined.head()

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0       1         1        NaN        NaN       NaN       NaN         6   
1       1         2        NaN        NaN       NaN       NaN         6   
2       2         1        NaN        NaN       NaN       NaN         5   
3       2         2        NaN        NaN       NaN       NaN         6   
4       2         3        NaN        NaN       NaN       NaN         6   

   pregend2  nbrnaliv  multbrth     ...      pubassis_i_r    basewgt_r  \
0       NaN         1       NaN     ...                 0  3410.389399   
1       NaN         1       NaN     ...                 0  3410.389399   
2       NaN         3         5     ...                 0  7226.301740   
3       NaN         1       NaN     ...                 0  7226.301740   
4       NaN         1       NaN     ...                 0  7226.301740   

   adj_mod_basewgt_r    finalwgt_r  secu_r  sest_r  cmintvw_r  cmlstyr  \
0        3869.349602   6448.271112       2       9       1231     1219   
1        3869.349602   6448.271112       2       9       1231     1219   
2        8567.549110  12999.542264       2      12       1231     1219   
3        8567.549110  12999.542264       2      12       1231     1219   
4        8567.549110  12999.542264       2      12       1231     1219   

   screentime   intvlngth  
0    19:56:43   67.563833  
1    19:56:43   67.563833  
2    14:54:03  106.018167  
3    14:54:03  106.018167  
4    14:54:03  106.018167  

[5 rows x 3331 columns]

In [34]:
# Following code taken from ThinkStats and annotated to show understanding
t = []
for col in joined.columns:
    try:
        # Skip any columns with tiny variance since they won't tell us much
        if joined[col].var() < 1e-7:
            continue

        # Create a model to examine how the baby's weight is related to 
        # the mother's age and the current column
        formula = 'totalwgt_lb ~ agepreg + ' + col
        model = smf.ols(formula, data=joined)
        
        # If less than half of the rows were used, skip this column
        if model.nobs < len(joined)/2:
            continue

        results = model.fit()
    except (ValueError, TypeError):
        continue

    t.append((results.rsquared, col))

# Sort the results by R-squared and display the top 30
t.sort(reverse=True)
for mse, name in t[:30]:
    print(name, mse)

('totalwgt_lb', 1.0)
(u'birthwgt_lb', 0.94981273059780091)
(u'lbw1', 0.30082407844707948)
(u'prglngth', 0.13012519488625029)
(u'wksgest', 0.12340041363360821)
(u'agecon', 0.1020314992815573)
(u'mosgest', 0.027144274639580579)
(u'babysex', 0.018550925293941312)
(u'race_r', 0.016199503586251329)
(u'race', 0.016199503586251329)
(u'nbrnaliv', 0.016017752709791666)
(u'paydu', 0.014003795578110934)
(u'rmarout03', 0.013430066465712986)
(u'birthwgt_oz', 0.013102457615710938)
(u'anynurse', 0.012529022541812429)
(u'bfeedwks', 0.012193688404499081)
(u'totincr', 0.011870069031174046)
(u'marout03', 0.011807801994371481)
(u'marcon03', 0.011752599354397653)
(u'cebow', 0.011437770919637935)
(u'rmarout01', 0.011407737138643737)
(u'rmarout6', 0.011354138472803199)
(u'marout01', 0.011269357246801448)
(u'hisprace_r', 0.011238349302033601)
(u'hisprace', 0.011238349302033601)
(u'mar1diss', 0.010961563590753065)
(u'fmarcon5', 0.0106049646843005)
(u'rmarout02', 0.010546913206563868)
(u'marcon02', 0.0104814017

The top three columns are related to the weight of the baby, so we obviously can't use them in our prediction.  Three of the next four columns are related to the length of the pregnancy which is also unknown until after the fact.  The other column is the mother's age at conception, but this seems redundant since we're already using the mother's age at the end of the pregnancy.  The sex of the baby is pretty high on the list and is most likely known by the 30th week of the pregnancy.  Race is the next highest factor, but as the reading mentions, can sometimes be coupled with various socioeconomic factors.  Out of the remaining columns, number of babies in the pregnancy, whether the mother owns or rents a home, and the mother's total family income could be known or reasonably estimated.

In [35]:
# Model from ThinkStats
formula = ('totalwgt_lb ~ agepreg + C(race) + babysex==1 + '
               'nbrnaliv>1 + paydu==1 + totincr')
results = smf.ols(formula, data=joined).fit()
print results.rsquared

# My modified model
formula = ('totalwgt_lb ~ agepreg + C(race) + babysex==1 + '
               'C(nbrnaliv) + paydu==1 + totincr')
results = smf.ols(formula, data=joined).fit()
print results.rsquared

0.0599863083304
0.0618127968117


Taking the model that Allen described in _ThinkStats_, I decided to make a slight modification regarding how multiple babies were handled.  My hypothesis was that I would be able to get better results by looking at this categorically (how many babies are in the pregnancy?) rather than as a boolean (is there more than one baby in the pregnancy?).  My reasoning is that it seems to me that triplets would each weigh less than twins who would each weigh less than a single child, and the specific number of babies in the pregnancy would probably be known.  The slight improvment in R$^2$ value seems to indicate that being more specific about this factor would improve your odds in guessing the weight of the baby/babies (though it's still not that much of an improvement over guessing without taking these factors into account).

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [20]:
import nsfg
import statsmodels.formula.api as smf
import numpy as np

df = nsfg.ReadFemPreg()

In [21]:
#Cleaning data
#converting all babies value (not ascertained (value 7), don't know (value 9)) and female (2)) 
#to 0 except male babies (value 1)
df.loc[df['babysex'] != 1, 'babysex'] = 0 

baby = df[df.outcome == 1 & (df.prglngth >= 30)]

In [22]:
"""
def r2 (column_no):
    try:
        variable = str(baby.columns[column_no])
        formula ='prglngth ~' + variable
        model = smf.ols(formula, data = baby)
        result = model.fit()
        return (result.rsquared, result.f_pvalue)
    except: 
        return (0,0)


variable_r2 = np.empty(len(baby.columns), dtype=object)
variable_p = np.empty(len(baby.columns), dtype=object)

for i in range(0, len(baby.columns)-1):
    variable_r2[i], variable_p[i] = r2(i)
    print baby.columns[i], variable_r2[i], variable_p[i]
"""

"\ndef r2 (column_no):\n    try:\n        variable = str(baby.columns[column_no])\n        formula ='prglngth ~' + variable\n        model = smf.ols(formula, data = baby)\n        result = model.fit()\n        return (result.rsquared, result.f_pvalue)\n    except: \n        return (0,0)\n\n\nvariable_r2 = np.empty(len(baby.columns), dtype=object)\nvariable_p = np.empty(len(baby.columns), dtype=object)\n\nfor i in range(0, len(baby.columns)-1):\n    variable_r2[i], variable_p[i] = r2(i)\n    print baby.columns[i], variable_r2[i], variable_p[i]\n"

In [23]:
import chap01soln
resp = chap01soln.ReadFemResp()
resp.index = resp.caseid
join = baby.join(resp, on='caseid', rsuffix='_r')
    
t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7:
            continue

        formula = ('prglngth ~' + name).encode('ascii')
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue

        results = model.fit()
        
        if name == 'babysex':
            print results.rsquared
            
    except (ValueError, TypeError):
        continue

    t.append((results.rsquared, name))

0.000193556647438


In [24]:
t.sort(reverse=True)
for mse, name in t[:30]:
    print name, mse

prglngth 1.0
wksgest 0.84834444419
totalwgt_lb 0.179548090415
lbw1 0.172861427756
birthwgt_lb 0.171898841283
mosgest 0.121330423471
prglngth_i 0.0152701066371
nbrnaliv 0.00769581202714
canhaver 0.00593856647669
datcon01_i 0.00430171280399
pregend1 0.00391214836732
con1mar1_i 0.00389093115987
anynurse 0.0033435395396
bfeedwks 0.00331167840872
rmarout08_i 0.00267785152974
marout08_i 0.00267785152974
coh1dur 0.00263088703117
datend02_i 0.00251119531636
datcon02_i 0.00251119531636
bpa_bdscheck1 0.00249843519185
gestasun_w 0.0024438427747
everhyst 0.00233317151081
totincr 0.00230344114738
agecon02_i 0.00230286994584
hyst 0.00229607320666
wantbold 0.00227411957336
ageprg02_i 0.002245700018
p1yrage 0.00220709648186
mar1con1_i 0.00219110817676
marcon08_i 0.00218870780021


Sorting out columns which gave r^2 > 0.0025,

In [25]:
sig_variable = []
unknown_list = ['prglngth', 'wksgest', 'totalwgt_lb', 'birthwgt_lb', 'lbw1, mosgest', 
                'prglngth_i', 'canhaver', 'lbw1', 'mosgest', 'nbrnaliv', 'anynurse',
                'bfeedwks', 'pregend1', 'rmarout11_i', 'datend02_i', 'marout11_i']
for mse, name in t[:30]:
    if (mse > 0.0025) & (name not in unknown_list):
        sig_variable.append ((mse, name))


for mse, name in sig_variable[:30]:
    print name, mse

datcon01_i 0.00430171280399
con1mar1_i 0.00389093115987
rmarout08_i 0.00267785152974
marout08_i 0.00267785152974
coh1dur 0.00263088703117
datcon02_i 0.00251119531636


datcon01_i: DATCON01 (CM date when pregnancy began - 1st ) Imputation Flag  

mar1con1_i: MAR1CON1 (Months btw/1st marriage & 1st conception (or interview)) Imputation Flag

marcon11_i: MARCON11 (Formal marital status when pregnancy began - 11th ) Imputation Flag 

datcon02_i: DATCON02 (CM date when pregnancy began - 2nd) Imputation Flag

cmlastlb: CM for R's most recent live birth

Removing variables with less significant r^2 values and information the co-workers cannot know beforehand, the variables we are left with are mainly imputation flags which are not good indicators to base the prediction on. Also, they are not variables which the co-worker might know.
Guessing might be a better option.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [5]:
import nsfg
import chap01soln
import statsmodels.formula.api as smf

df = nsfg.ReadFemPreg()
df = df[df.prglngth>30]
resp = chap01soln.ReadFemResp()
resp.index = resp.caseid
join = df.join(resp, on='caseid', rsuffix='_r')

t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7:
            continue
        formula = 'prglngth ~ agepreg + ' + str(name)
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    t.append((results.rsquared, name))
    
t.sort(reverse=True)
for mse, name in t[:30]:
    print name, mse

prglngth 1.0
wksgest 0.810466988405
agecon 0.781350418865
totalwgt_lb 0.125503308033
birthwgt_lb 0.120679453558
lbw1 0.103838874743
mosgest 0.0982884576401
prglngth_i 0.0215849926795
pregend1 0.00843280219463
outcome 0.00787817147659
prgoutcome 0.0078466125794
canhaver 0.00675998334425
datcon01_i 0.00590185285861
con1mar1_i 0.00559512377644
nbrnaliv 0.00458014116379
mar1con1_i 0.00328669224476
anynurse 0.00291848398128
bfeedwks 0.00256701215881
evuseint 0.00216793828297
rmarout11_i 0.00216747068096
marout11_i 0.00216747068096
marcon11_i 0.00216747068096
datend02_i 0.00207103036161
datcon02_i 0.00207103036161
agecon02_i 0.00202887790059
ageprg02_i 0.00199128042072
paydu 0.00198269210293
fmarcon5_i 0.00194217103912
cmlastlb_r 0.00190168382203
cmlastlb 0.00190168382203


Only some of the variables mentioned above have values/ approx. values associated with them. I will probably first find all the information I can get and use those variables. My confidence on the following variables will be the most: agecon (age at time of conception) because the mother's age will differ at most my a month or two if the baby is late or early and canhaver (ability to have another baby).

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

We've seen a link between the pregnancy length and how many pregnancies a woman has had, so that could be taken into account. We could also use her age.

I'm not entirely sure what you want with this question. Code? More analysis?

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

Well the variables I would use, pulling from the information drawn in the nsfc, would be first, whether or not this baby is my coworker's first pregnancy or has she had multiple pregnancies. As we know from reading the majority of thinkstats2, first borns come out a little sooner than later borns. Additionally I would also use babysex because boys are heavier than girls. If so, then heavier babies might be sooner about the pregnancy day. Then I would ask if her family had a tendency of having twins or more than one baby at a time. I would use the nbraliv to use this quality to include in my prediction model. I would also include age of pregancy, how old my coworker friend is at the time of her pregnancy.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

<p>
    If we look at variables related to date of birth and known before birth from our previous explorations of the data set: 
</p>

In [7]:
import nsfg
import thinkplot 
import numpy as np
import pandas as pd
import thinkstats2
from sklearn import cross_validation
from sklearn.linear_model import LogisticRegression
from sklearn import cross_validation
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import SelectKBest, f_classif
%matplotlib inline

pregs = nsfg.ReadFemPreg()
print(pregs.describe)



<bound method DataFrame.describe of        caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0           1         1        NaN        NaN       NaN       NaN         6   
1           1         2        NaN        NaN       NaN       NaN         6   
2           2         1        NaN        NaN       NaN       NaN         5   
3           2         2        NaN        NaN       NaN       NaN         6   
4           2         3        NaN        NaN       NaN       NaN         6   
5           6         1        NaN        NaN       NaN       NaN         6   
6           6         2        NaN        NaN       NaN       NaN         6   
7           6         3        NaN        NaN       NaN       NaN         6   
8           7         1        NaN        NaN       NaN       NaN         5   
9           7         2        NaN        NaN       NaN       NaN         5   
10         12         1        NaN        NaN       NaN       NaN         5   
11         14   

<p>
    Variables usable would be - gender of the baby, current weight of baby vs a standardized, expected weight along with growth data, race of mother vs average, data from previous siblings.  
</p>

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

Some of the variables that are known would be Age of mother, race, firstchild, length of pregnancy for previous children.

In [2]:
import nsfg
import chap01soln
import statsmodels.formula.api as smf

df = nsfg.ReadFemPreg()
resp = chap01soln.ReadFemResp()

We will focus on respondent with live births so outcome is true and the pregnancy was longer than 30 weeks

In [3]:
live = df[(df.outcome ==1) &(df.prglngth >30)]

resp.index = resp.caseid
join = live.join(resp, on='caseid', rsuffix='_r')

In [4]:
t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7:
            continue
        formula = ('prglngth ~ ' + name).encode('ascii')
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    t.append((results.rsquared, name))

In [5]:
t.sort(reverse=True)
for mse, name in t[:30]:
    print name, mse

prglngth 1.0
wksgest 0.806243411614
totalwgt_lb 0.124457431481
birthwgt_lb 0.119773078049
lbw1 0.103725422046
mosgest 0.0956243198959
prglngth_i 0.0220537757965
canhaver 0.00605049526819
datcon01_i 0.00581775529989
con1mar1_i 0.00554637613623
nbrnaliv 0.00457756578553
mar1con1_i 0.00315080225387
anynurse 0.00245202488371
bfeedwks 0.00236918394467
pregend1 0.0022493894338
rmarout11_i 0.00224362796811
marout11_i 0.00224362796811
marcon11_i 0.00224362796811
cmlastlb_r 0.0020431424422
cmlastlb 0.0020431424422
datend02_i 0.00201248339274
datcon02_i 0.00201248339274
agecon02_i 0.00198828676887
fmarcon5_i 0.00196815932426
ageprg02_i 0.00195280723688
evuseint 0.00189175277586
paydu 0.001876821903
marcon03_i 0.00179114104015
anymschp 0.00177999843325
gestasun_m 0.00165713195503


The obvious correlation is with prglngth but that is not an information that is available to the co-workers. None of the variables presented make a good case for guessing I would say.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [11]:
import statsmodels.formula.api as smf
import numpy as np 
import pandas as pd
import nsfg

preg = nsfg.ReadFemPreg()
live = preg[preg.outcome == 1]
live_30 = live[live.prglngth >= 30]
firsts = live_30[live_30.birthord == 1]
others = live_30[live_30.birthord != 1]

## Variables can be used for the prediction


There are many important data that might be useful in prediciting one's baby's date of birth. The age of pregnance might have some influence on date of birth: young mother and others may have different length of pregnancy. The birth order also matters. First baby might be a little different in length of pregnancy than the others. Thus, I will use agepreg and birthord to predict the date of birth. I also will need to know the specific date of pregnancy in order to make the bet on a specific week.

Also, since it is already the 30th week of pregnancy, it is probably good to only select data of pregnancy length longer than 30 weeks among the live birth group.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

You could use mother's age, nth-child, and the baby's sex, then compare them to the pregnancy length. Using R^2 values and p-values, you could make a best guesstimate.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [5]:
data = nsfg.ReadFemPreg()
live = data[data.prglngth>30]
resp = chap01soln.ReadFemResp()
resp.index = resp.caseid
join = live.join(resp, on = "caseid", rsuffix = "_r")
t = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7:
            continue
        formula = 'prglngth ~ agepreg + ' + str(name)
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    t.append((results.rsquared, name))
    
t.sort(reverse=True)
for mse, name in t[:30]:
    print name, mse

prglngth 1.0
wksgest 0.810466988405
agecon 0.781350418865
totalwgt_lb 0.125503308033
birthwgt_lb 0.120679453558
lbw1 0.103838874743
mosgest 0.0982884576401
prglngth_i 0.0215849926795
pregend1 0.00843280219459
outcome 0.00787817147656
prgoutcome 0.00784661257936
canhaver 0.00675998334427
datcon01_i 0.00590185285858
con1mar1_i 0.00559512377641
nbrnaliv 0.00458014116394
mar1con1_i 0.00328669224473
anynurse 0.00291848398136
bfeedwks 0.00256701215878
evuseint 0.0021679382829
rmarout11_i 0.00216747068093
marout11_i 0.00216747068093
marcon11_i 0.00216747068093
datend02_i 0.00207103036159
datcon02_i 0.00207103036159
agecon02_i 0.00202887790055
ageprg02_i 0.00199128042069
paydu 0.00198269210289
fmarcon5_i 0.00194217103909
cmlastlb_r 0.00190168382206
cmlastlb 0.00190168382206


Using the code from the text book, we can see the top factors contributing to pregnancy length are listed as above.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [30]:
# Import modules and libraries
import first
import statsmodels.formula.api as smf

# Create 3 df, one for live baby, firsts, others
# Because we are already at 30 weeks, we will only take data
# from the live where the baby is over 30 weeks old.
live, firsts, others = first.MakeFrames()
df = live[live.prglngth>30]


formula = ('prglngth ~ birthord + race + nbrnaliv + babysex')
model = smf.ols(formula, data=df)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               prglngth   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     13.35
Date:                Thu, 28 Jan 2016   Prob (F-statistic):           7.54e-11
Time:                        21:13:14   Log-Likelihood:                -18261.
No. Observations:                8878   AIC:                         3.653e+04
Df Residuals:                    8873   BIC:                         3.657e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept     39.6947      0.152    260.759      0.000        39.396    39.993
birthord      -0.0563      0.019     -2.894      0.004        -0.094    -0.018
race           0.0476      0.036      1.340      0.180        -0.022     0.117
nbrnaliv      -0.7054      0.114     -6.187      0.000        -0.929    -0.482
babysex       -0.0546      0.040     -1.360      0.174        -0.133     0.024
==============================================================================
Omnibus:                     1579.495   Durbin-Watson:                   1.622
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6065.294
Skew:                          -0.851   Prob(JB):                         0.00
Kurtosis:                       6.674   Cond. No.                         30.3
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Observing the major variables mentioned in the reading pre-birth, which are birthord, race, nbrnaliv, and babysex, it seems as if race has the greatest coefficient. All the other coefficients are negative, indicating that they are either in the negative direction or the prediction is worse than random. However, magnitude wise, nbrnaliv is the greatest, which indicates that the number of children born at once heavily effect the birth length.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

#### Response

As explained in the reading, to make the best prediction, you could use

 babysex  - Is the baby going to be a boy or girl? (If known)
 nbrnaliv - Is the mother expecting twins or triplets?
 paydu    - Does the mother own a home?
 totincr  - What is her total income? (As co=workers we can likely estimate)
 agepreg  - How old is our co-worker?
 

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

If I were to assume that I worked at a place where more than 95% of my coworkers earned more than \$75,000, I would probably use whether the baby was male or female, whether the mother will be birthing twins or multiplets, and whether or not they already were comfortably living in a house before the 30th week of the pregnancy. While these variables do not help my chances much, they seem to be the most likely candidates based my intuition that stress may also play a role (hence the moving into a house or not) but cannot be accurately measured with a survey.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

* Age of mother
* Race
* Is the first child
* If not first child, lengh of pregnancy for previous child/children

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

______
I need to first gather data, as well as only include the columns that happen at the 30 week pregnancy length and longer

In [43]:
import numpy as np
import thinkstats2
import thinkplot
import nsfg
import statsmodels.formula.api as smf

df = nsfg.ReadFemPreg()
df = df[df.prglngth>=30]
df

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0           1         1        NaN        NaN       NaN       NaN         6   
1           1         2        NaN        NaN       NaN       NaN         6   
2           2         1        NaN        NaN       NaN       NaN         5   
3           2         2        NaN        NaN       NaN       NaN         6   
4           2         3        NaN        NaN       NaN       NaN         6   
5           6         1        NaN        NaN       NaN       NaN         6   
6           6         2        NaN        NaN       NaN       NaN         6   
7           6         3        NaN        NaN       NaN       NaN         6   
8           7         1        NaN        NaN       NaN       NaN         5   
9           7         2        NaN        NaN       NaN       NaN         5   
10         12         1        NaN        NaN       NaN       NaN         5   
11         14         1        NaN        NaN       NaN       NaN         6   
12         14         2        NaN        NaN       NaN       NaN         6   
15         15         2        NaN        NaN       NaN       NaN         6   
16         15         3        NaN        NaN       NaN       NaN         6   
17         18         1        NaN        NaN       NaN       NaN         5   
19         21         1        NaN        NaN       NaN       NaN         6   
20         21         2        NaN        NaN       NaN       NaN         6   
21         23         1        NaN        NaN       NaN       NaN         6   
23         24         1        NaN        NaN       NaN       NaN         6   
24         24         2        NaN        NaN       NaN       NaN         6   
25         24         3        NaN        NaN       NaN       NaN         6   
26         28         1        NaN        NaN       NaN       NaN         6   
27         31         1        NaN        NaN       NaN       NaN         6   
28         31         2        NaN        NaN       NaN       NaN         6   
29         31         3        NaN        NaN       NaN       NaN         6   
31         36         2        NaN        NaN       NaN       NaN         6   
32         38         1        NaN        NaN       NaN       NaN         5   
33         38         2        NaN        NaN       NaN       NaN         6   
34         38         3        NaN        NaN       NaN       NaN         6   
...       ...       ...        ...        ...       ...       ...       ...   
13551   12541         1        NaN        NaN       NaN       NaN         6   
13552   12541         2        NaN        NaN       NaN       NaN         6   
13553   12541         3        NaN        NaN       NaN       NaN         6   
13554   12541         4        NaN        NaN       NaN       NaN         6   
13555   12544         1        NaN        NaN       NaN       NaN         5   
13556   12544         2        NaN        NaN       NaN       NaN         5   
13557   12544         3        NaN        NaN       NaN       NaN         5   
13559   12545         2        NaN        NaN       NaN       NaN         6   
13560   12546         1        NaN        NaN       NaN       NaN         6   
13561   12546         2        NaN        NaN       NaN       NaN         6   
13562   12547         1        NaN        NaN       NaN       NaN         6   
13563   12547         2        NaN        NaN       NaN       NaN         6   
13564   12547         3        NaN        NaN       NaN       NaN         6   
13565   12550         1        NaN        NaN       NaN       NaN         6   
13566   12551         1        NaN        NaN       NaN       NaN         5   
13569   12556         1        NaN        NaN       NaN       NaN         6   
13570   12556         2        NaN        NaN       NaN       NaN         6   
13571   12556         3        NaN        NaN       NaN       NaN         6   
13572   12556         4        NaN        NaN       NaN  

I'm going to split up the data to only include the variables that I think could possibly make an impact on the pregnancy length, which I decided are age, race, birth order, and weight. I made a new dataframe that only contains this data as well as the actual pregnancy lenght to train. I also dropped the NA values.

In [44]:
df = df.dropna(subset=['prglngth', 'agepreg', 'race', 'birthord', 'basewgt'])
newdf = df[['prglngth', 'agepreg', 'race', 'birthord', 'basewgt']].copy()

lengths = df.prglngth
ages = df.agepreg
races = df.race
orders = df.birthord
weights = df.basewgt

newdf.groupby('prglngth').mean()


agepreg      race  birthord       basewgt
prglngth                                             
30        23.308261  1.630435  1.768116   4406.032576
31        24.595185  1.888889  1.592593   3575.806203
32        24.781217  1.756522  1.860870   4384.832520
33        24.892857  1.653061  1.836735   4532.070787
34        24.547500  1.750000  1.783333   5229.789681
35        24.524630  1.868167  1.749196   4061.930607
36        25.136947  1.803738  1.760125   4217.903898
37        25.598308  1.848352  1.850549   4271.932559
38        26.099918  1.894563  1.805601   4134.976981
39        24.717903  1.797784  1.906243   4387.425108
40        25.454704  1.873656  1.793907   4402.155208
41        24.851465  1.877342  1.555366   4268.820754
42        24.402774  1.887195  1.582317   4552.579632
43        25.281149  1.871622  1.601351   4255.990287
44        24.957391  1.760870  1.652174   4922.211229
45        26.688000  1.600000  1.400000   2192.138603
46        15.410000  1.000000  1.000000   3613.100749
47        18.660000  1.000000  1.000000   1914.130645
48        27.342857  1.714286  1.857143   2132.568260
50        21.785000  1.500000  2.000000  12215.367496

I started by comparing the dependent variable, pregnancy length, with age pregnant

In [45]:
formula1 = 'prglngth ~ agepreg'
model = smf.ols(formula1, data=newdf)
results = model.fit()
print results.summary()


                            OLS Regression Results                            
Dep. Variable:               prglngth   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     1.958
Date:                Thu, 28 Jan 2016   Prob (F-statistic):              0.162
Time:                        18:48:35   Log-Likelihood:                -19815.
No. Observations:                9022   AIC:                         3.963e+04
Df Residuals:                    9020   BIC:                         3.965e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept     38.5986      0.105    366.825      0.0

I then compared the dependent variable, pregnancy length, with race

In [46]:
formula1 = 'prglngth ~ race'
model = smf.ols(formula1, data=newdf)
results = model.fit()
print results.summary()

                            OLS Regression Results                            
Dep. Variable:               prglngth   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     11.99
Date:                Thu, 28 Jan 2016   Prob (F-statistic):           0.000538
Time:                        18:48:35   Log-Likelihood:                -19810.
No. Observations:                9022   AIC:                         3.962e+04
Df Residuals:                    9020   BIC:                         3.964e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept     38.4880      0.077    500.192      0.0

I then compared the dependent variable, pregnancy length, with birth order of the child

In [47]:
formula1 = 'prglngth ~ birthord'
model = smf.ols(formula1, data=newdf)
results = model.fit()
print results.summary()

                            OLS Regression Results                            
Dep. Variable:               prglngth   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     6.474
Date:                Thu, 28 Jan 2016   Prob (F-statistic):             0.0110
Time:                        18:48:35   Log-Likelihood:                -19812.
No. Observations:                9022   AIC:                         3.963e+04
Df Residuals:                    9020   BIC:                         3.964e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept     38.8447      0.046    839.111      0.0

I then compared the dependent variable, pregnancy length, with respondents weight

In [48]:
formula1 = 'prglngth ~ basewgt'
model = smf.ols(formula1, data=newdf)
results = model.fit()
print results.summary()

                            OLS Regression Results                            
Dep. Variable:               prglngth   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                    0.4532
Date:                Thu, 28 Jan 2016   Prob (F-statistic):              0.501
Time:                        18:48:35   Log-Likelihood:                -19815.
No. Observations:                9022   AIC:                         3.963e+04
Df Residuals:                    9020   BIC:                         3.965e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept     38.7255      0.034   1142.881      0.0

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [86]:
#set up the necessary imports
import nsfg
import chap01soln
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import statsmodels.formula.api as smf

df = nsfg.ReadFemPreg()
resp = chap01soln.ReadFemResp()

In [87]:
#let's just focus on the respondent data with live births (since we'll assume the best)
#also assume within a month and change before an "average" length pregnancy of 40 weeks is when the pool occurs
#thus we know at last some length of the pregnancy. We also know the age of our pregnant friend.

live = df[df.outcome == 1]
live = live[live.prglngth > 30]

resp.index = resp.caseid
join = live.join(resp, on='caseid', rsuffix='_r')

In [88]:
#we want to test on a variety of factors to determine what relationships might exist
test_table = []
for name in join.columns:
    try:
        if join[name].var() < 1e-7: #make sure that there is a little but of variance
            continue
            
        formula = 'prglngth ~ agecon + ' + str(name) #predicting prglngth as a function of age, other parameter
        model = smf.ols(formula, data=join)
        if model.nobs < len(join)/2:
            continue
        results = model.fit()
    except (ValueError, TypeError):
        continue
    test_table.append((results.rsquared, name))

In [89]:
test_table.sort(reverse=True) #sort by highest value
for mse, name in test_table[:30]:
    print(name, mse) #print the top ten

(u'prglngth', 1.0)
(u'wksgest', 0.80647560229554416)
(u'agepreg', 0.77722780709056005)
('totalwgt_lb', 0.12589504948051811)
(u'birthwgt_lb', 0.12104803951300569)
(u'lbw1', 0.10399010733882053)
(u'mosgest', 0.095653948252776844)
(u'prglngth_i', 0.022702898461402943)
(u'canhaver', 0.0060528525930908517)
(u'datcon01_i', 0.006021881275759422)
(u'con1mar1_i', 0.0057470796692380421)
(u'nbrnaliv', 0.0046402822824620493)
(u'mar1con1_i', 0.0032946548980058443)
(u'anynurse', 0.0032494403967081587)
(u'bfeedwks', 0.0027926146173661293)
(u'agebaby1', 0.0027862747948664834)
(u'rmarout11_i', 0.0023613798588607571)
(u'marout11_i', 0.0023613798588607571)
(u'marcon11_i', 0.0023613798588607571)
(u'paydu', 0.0023452835507742353)
(u'pregend1', 0.0022912806602890523)
(u'datend02_i', 0.0021324893643125398)
(u'datcon02_i', 0.0021324893643125398)
(u'cmlastlb_r', 0.0021148239919197565)
(u'cmlastlb', 0.0021148239919197565)
(u'agecon02_i', 0.0021129060195237415)
(u'ageprg02_i', 0.0020748312204004193)
(u'fmarcon5_

Based on this information, I believe that the more useful parameters (which would be either known or guessable by the folks participating in the pool) would/may be: canhavr (perhaps if she is very close within people in the office), con1mar1 (the months between conception and marriage), mar1con1 (months between marriage and conception), agebaby1 (age at first birth, assuming more than one child), nbrnaliv (multiple births, if known), and paydu (current living quarters). I will say that none of these, without further investigation, would make a compelling case on their own, but together their might be something that can be predicted better than random guessing.

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

There are several variables I would consider for the best prediction model. Certain variables that are already known are the "race", "first child" and "age of the mother". In addition to that, we know how long the pregnancy for the earlier kids lasted. Now below, one can see the main variables that I would consider:
1. A boolean value that denotes if the child is a first child or not. 
2. As afore-mentioned, race apparently seemed to affect the babies' weights
3. Total number of babies born by a mother 
4. Financial comfort variables -- income and its effect on number and duration of pregnancies
5. Height and weight of the child 
6. General health of the mother
7. Gender of baby - difficult to distinguish until further into the pregnancy (?)

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

In [2]:
"""babysex, nbrnaliv, paydu, totincr, race are all the variables I'd look at, based on information given in the reading""" 

"babysex, nbrnaliv, paydu, totincr, race are all the variables I'd look at, based on information given in the reading"

## Question 1

Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool. 

Your co-worker's age

Whether or not it is your co-worker's first baby

Your co-worker's race

The sex of your co-worker's baby

The season your co-worker's baby was concieved in

How many babies your co-worker is having

## Clarifying Questions

Use this space to ask questions regarding the content covered in the reading. These questions should be restricted to helping you better understand the material. For questions that push beyond what is in the reading, use the next answer field. If you don't have a fully formed question, but are generally having a difficult time with a topic, you can indicate that here as well.

It seems like there are a lot of variables that Allen defined earlier on in think stats that he used in this chapter. It wasn't immediately ovious to me (maybe I missed something) in this chapter or the previous one how these variables would be defined, and it makes reading the chapters/doing the exercises out of order more confusing/time consuming. 

One of the questions that I had when trying to answer a question like this is whether I use only data of women 30 weeks pregnant to answer this question, or is it acceptable to use all of the data? Or only data for women pregnant less than 30 weeks? More than 30 weeks? How do I know what the "right" portion of the data to answer a question like this? 

I don't entirely understand the code examples in the chapter (that I used to solve this exercise) in the sense that I wouldn't know when to use it, or how/why these particular things are done. 

It seems like a lot of factor can be coupled, such as home ownership status and annual income.  Are there any pitfalls to be aware of regarding this?  Does it matter at all when doing regressions?

So I heavily used ThinkStats to solve this problem. I am not sure if we are supposed to do that or get some basic understanding from that and write our own models?

How is this questions related to the reading? We aren't creating a model unless the point is to find desirable features to include in the model

In a case like this what would the model entail? : I'm confused to how it would learn. There is a right and wrong answer but where does the prediction come from? In the case of the titanic it is yes or no with only two options: here there are multiple prediction dates that could be wrong and one that is right. Does it become a model that gives likilihood on the chances of each extra week then?

I am not sure, when we set the formula for the smf.ols(), what the == or < do. (ex birthord == 1). I read that they make it a boolean but they seem to alter the coefficient values printed by summary. 

I'm still not sure about how to find the end goal here. I did some observations on different statitics on the possibly affecting variables, but am not sure how to visualize or come to a conclusion about the information.

I just don't know if my intuition was really built well during this chapter for models like this, and wonder if we could walk through formula invocation could be reviewed in class further.

I am not entirely clear about the part where Allen joins the two NSFG files. I understand why he did it but I do not understand why he can't concatenate the race field into one, if you had just cleaned up the data in one of the tables to match the other one before joining. Is this something that is not possible?

I am not also not sure at what point an estimated slope or intercept becomes statistically significant.

When you have a dataset that's as large as the one we have, how do you go about managing it/selecting what columns to use for analysis? There are more than 1000 columns and looking through them all seems a little impratical.

How do we scan all 244 variables and get the correlation values for all of them in a nice list? Is it some data mining technique, or do we have to filter them and test each independently?

Overall I had great struggle with this chapter as I could not really understand what to do other than filtering out the dataframe for pregnancies over 30 weeks long. I've worked with a few people and the overall understanding of this chapter compared to the other ones seemed a bit lower. Could we possibly go over this chapter in a little more depth in class?

## Additional Resources / Explorations

If you found any useful resources, or tried some useful exercises that you'd like to report please do so here. Let us know what you did, what you learned, and how others can replicate it.

<p> i think that there was too much work for these two days: 2 notebooks + a project step was too confusing. </p>

I found [this](http://www.cdc.gov/nchs/data/nsfg/NSFG_2006-2010_UG_App1a_FemRespFileIndex.pdf) a useful resource in addition to the codebook.